### PR TITLE
Add upstream patch to allow building on aarch64-apple-darwin

### LINKED ISSUE #72: Supporting aarch64-apple-darwin
I was hoping that OpenSSL would accept [the PR that adds support to their build system for aarch64-apple-darwin](https://github.com/openssl/openssl/pull/12369), but with 1.1.1h released without it, it seems unlikely by the end of 2020. As such, I'd like to discuss how the project feels about **applying our own patches** on top of the official release. 

Specifically, the patch in question only modifies the build system to use existing pieces in new ways.

With that applied, and a single line added to this crate to map target triples, it's possible to cross-compile to aarch64-apple-darwin:

```
% SDKROOT=$(xcrun -sdk macosx11.0 --show-sdk-path) MACOSX_DEPLOYMENT_TARGET=$(xcrun -sdk macosx11.0 --show-sdk-platform-version) cargo +462ec6baef52edea591eab1a070ed4e18b80d094 build --target=aarch64-apple-darwin --features vendored
   Compiling cc v1.0.60
   Compiling pkg-config v0.3.18
   Compiling autocfg v1.0.1
   Compiling libc v0.2.77
   Compiling bitflags v1.2.1
   Compiling foreign-types-shared v0.1.1
   Compiling openssl v0.10.30 (/Users/shep/Projects/rust-openssl/openssl)
   Compiling cfg-if v0.1.10
   Compiling lazy_static v1.4.0
   Compiling foreign-types v0.3.2
   Compiling openssl-src v111.10.2+1.1.1g (/Users/shep/Projects/openssl-src-rs)
   Compiling openssl-sys v0.9.58 (/Users/shep/Projects/rust-openssl/openssl-sys)
    Finished dev [unoptimized + debuginfo] target(s) in 29.71s
```

This should make it possible to build Cargo for this platform as well.

### PR DESCRIPTION
**Pull request by shepmaster**
Closes #72

### CODE REVIEW

**Comment by shepmaster** [comment]
With this branch, I was able to cross-compile Cargo:

```diff
diff --git a/Cargo.toml b/Cargo.toml
index aa6be2f3d..fb8986e0c 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -119,3 +119,6 @@ doc = false
 deny-warnings = []
 vendored-openssl = ["openssl/vendored"]
 pretty-env-logger = ["pretty_env_logger"]
+
+[patch.crates-io]
+openssl-src = { git = "https://github.com/shepmaster/openssl-src-rs",  branch = "aarch64-apple-darwin" }
```

```
% SDKROOT=$(xcrun -sdk macosx11.0 --show-sdk-path) MACOSX_DEPLOYMENT_TARGET=$(xcrun -sdk macosx11.0 --show-sdk-platform-version) cargo +462ec6baef52edea591eab1a070ed4e18b80d094 build --target=aarch64-apple-darwin --features vendored-openssl

    Finished dev [unoptimized + debuginfo] target(s) in 1m 34s
```

**Comment by alexcrichton** [comment]
Is it possible to test this on CI? Or do the macs on github actions not have the sdk installed?

**Comment by shepmaster** [comment]
> Is it possible to test this on CI

It’s *kind* of a catch 22 situation. This change is needed to be able to build Cargo for aarch64 in Rust CI. 

Right now, the only artifact you could use would be a random try build. With that, and some environment variables, you could cross compile to aarch64 on GitHub Actions or Azure Pipelines (haven’t looked at other CI providers). 

None of the CI providers have aarch64 Darwin hardware, so you can’t actually run any tests.

**Comment by shepmaster** [comment]
Once https://github.com/rust-lang/rust/pull/75991 is merged, then presumably those artifacts would be available to rustup in nightly, making it one step easier.

File: src/lib.rs [hunk 1]
```diff
@@ -467,6 +473,44 @@ fn apply_patches(target: &str, inner: &Path) {
...
+        cflags           => add("-arch arm64"),
+        lib_cppflags     => add("-DL_ENDIAN"),
+        bn_ops           => "SIXTY_FOUR_BIT_LONG",
+        perlasm_scheme   => "ios64",
+    },"#;
+
+    let x86_64_stanza = buf.find(r#"    "darwin64-x86_64-cc""#).unwrap();
+    buf.insert_str(x86_64_stanza, PATCH);
+
+    File::create(&path)
```
Comment (sfackler) [inline src/lib.rs:508]: This can just use `fs::write(&path, buf.as_bytes())`

File: src/lib.rs [hunk 2]
```diff
@@ -467,6 +473,44 @@ fn apply_patches(target: &str, inner: &Path) {
...
+        return;
+    }
+
+    // Apply build system changes to allow configuring and building
+    // for Apple's ARM64 platform.
+    // https://github.com/openssl/openssl/pull/12369
+
+    let mut buf = String::new();
+    let path = inner.join("Configurations/10-main.conf");
+    File::open(&path).unwrap().read_to_string(&mut buf).unwrap();
```
Comment (sfackler) [inline src/lib.rs:487]: This can use `fs::read_to_string`

File: src/lib.rs (hunk 1 above)
Comment (shepmaster) [reply src/lib.rs:508]: Ok, I just copied the existing patch structure and assumed that the library was striving to support older Rust versions. 

Would you like me to clean up the MUSL to use these functions as well?

Comment (sfackler) [reply src/lib.rs:508]: The oldest version we need to support is 1.31 (at least that's what rust-openssl tests against), so the fs:: methods should be fine there. Yeah, might as well clean up the MUSL one as well.

File: src/lib.rs [hunk 3]
```diff
@@ -467,6 +473,44 @@ fn apply_patches(target: &str, inner: &Path) {
...
+
+    assert!(
+        !buf.contains("darwin64-arm64-cc"),
+        "{} already contains instructions for aarch64-apple-darwin",
+        path.display(),
+    );
+
+    const PATCH: &'static str = r#"
```
Comment (est31) [inline src/lib.rs:495]: `'static` has not been required on constants [since 1.17.0](https://github.com/rust-lang/rust/blob/master/RELEASES.md#language-28)

Comment (shepmaster) [reply src/lib.rs]: Yup. Again, just being overly cautious because I wasn't aware of the crate's MSRV requirements.

**Comment by alexcrichton** [comment]
Ah right sorry forgot that we needed a Rust target. In that case this looks good to me, thanks again @shepmaster!

**Comment by shepmaster** [comment]
I’ll try to submit a CI PR once the target lands in nightly. I assume that testing via cross compiling and with nightly should be ok?

**Comment by shepmaster** [comment]
Thanks!

**Comment by alexcrichton** [comment]
Yeah that'd be great! And yeah I don't expect to run tests, but it would be good to exercise the patching and building logic if we can.

**Pull request merged**